clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/op/job/KubernetesRunJobOperation.java

```diff
@@ -0,0 +1,109 @@
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.*;
+
+@Slf4j
+public class KubernetesRunJobOperation implements AtomicOperation<DeploymentResult> {
+  private static final String OP_NAME = "RUN_KUBERNETES_JOB";
+  private final KubernetesRunJobOperationDescription description;
+  private final KubernetesV2Credentials credentials;
+  private final KubernetesResourcePropertyRegistry registry;
+  private final Namer namer;
+  private final ArtifactProvider provider;
+
+  public KubernetesRunJobOperation(KubernetesRunJobOperationDescription description,
+                                   KubernetesResourcePropertyRegistry registry,
+                                   ArtifactProvider provider) {
+    this.description = description;
+    this.credentials = (KubernetesV2Credentials) description.getCredentials().getCredentials();
+    this.registry = registry;
+    this.provider = provider;
+    this.namer = NamerRegistry.lookup()
+      .withProvider(KubernetesCloudProvider.getID())
+      .withAccount(description.getCredentials().getName())
+      .withResource(KubernetesManifest.class);
+  }
+
+  private static Task getTask() {
+    return TaskRepository.threadLocalTask.get();
+  }
+
+  public DeploymentResult operate(List _unused) {
+    getTask().updateStatus(OP_NAME, "Running Kubernetes job...");
+    KubernetesManifest jobSpec = this.description.getManifest();
+    KubernetesKind kind = jobSpec.getKind();
+    if (kind != KubernetesKind.JOB) {
+      throw new IllegalArgumentException("Only kind of Job is accepted for the V2 Run Job operation.");
+    }
+
+    if (jobSpec.get("metadata") == null) {
+      jobSpec.put("metadata", new HashMap<>());
+    }
+
+    if (!this.description.getNamespace().isEmpty()){
+      jobSpec.setNamespace(this.description.getNamespace());
+    }
+    // append a random string to the job name to avoid collision
+    String currentName = jobSpec.getName();
+    String postfix = Long.toHexString(Double.doubleToLongBits(Math.random()));
+    jobSpec.setName(currentName + "-" + postfix);
+
+    KubernetesDeployManifestDescription deployManifestDescription = new KubernetesDeployManifestDescription();
+    // setup description
+    List<KubernetesManifest> manifests = new ArrayList<>();
+    manifests.add(jobSpec);
+
+    Moniker moniker = new Moniker();
+    moniker.setApp(description.getApplication());
+
+    deployManifestDescription.setManifests(manifests);
+    deployManifestDescription.setSource(KubernetesDeployManifestDescription.Source.text);
+    deployManifestDescription.setMoniker(moniker);
+    deployManifestDescription.setCredentials(description.getCredentials());
+
+    KubernetesDeployManifestOperation deployManifestOperation = new KubernetesDeployManifestOperation(deployManifestDescription, registry, provider);
+    OperationResult operationResult = deployManifestOperation.operate(new ArrayList());
+    DeploymentResult deploymentResult = new DeploymentResult();
```

Comment:
I wonder if we could pull this logic out so that the run job stage first has a "generate manifest" task then, a "deploy manifest" task that deploys that.  Not sure exactly what that would look like, but maybe that would work?

Reply:
Perhaps. I feel like that would take a fair bit of refactoring in the `RunJobStage` to understand the difference between Kubernetes V1/V2 and Titus.

Reply:
Yeah, it's probably not worth the effort---the other things above about making separate classes were more obvious refactors, this one has less of an obvious benefit.